File: android/src/main/java/com/coinbase/waassdkreactnative/PoolService.java
package com.coinbase.waassdkreactnative;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;
import com.facebook.react.module.annotations.ReactModule;
import com.waassdkinternal.v1.Pool;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * React-native wrapper for {@link com.coinbase.waassdk.PoolService}
 */
@ReactModule(name = PoolService.NAME)
public class PoolService extends ReactContextBaseJavaModule {
  public static final String NAME = "PoolService";
  private String poolsErr = "E_POOL_SERVICE";

  private String uninitializedErr = "pool service must be initialized";

  ExecutorService executor;

  // The handle to the Go PoolService client.
  com.coinbase.waassdk.PoolService poolClient;

  PoolService(ReactApplicationContext reactContext) {
    super(reactContext);
    this.executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
  }

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }

  /**
   * Initializes the PoolService with the given Cloud API Key parameters. Resolves on success;
   * rejects with an error otherwise.
   */
  @ReactMethod
  public void initialize(String apiKeyName, String privateKey, Promise promise) {
    if (poolClient != null) {
      promise.resolve(true);
      return;
    }

    try {
      poolClient = new com.coinbase.waassdk.PoolService(apiKeyName, privateKey, executor);
      promise.resolve(null);
    } catch (Exception e) {
      promise.reject("initialize pool failed : ", e);
    }
  }

  /**
   * Creates a Pool with the given parameters.  Resolves with the created Pool object on success; rejects with an error
   * otherwise.
   */
  @ReactMethod
  public void createPool(String displayName, String poolID, Promise promise) {
    if (poolClient == null) {
      promise.reject(poolsErr, uninitializedErr);
      return;
    }

    WaasPromise.resolveMap(poolClient.createPool(displayName, poolID), promise, (Pool pool) -> {
      WritableMap outMap = new WritableNativeMap();
      outMap.putString("name", pool.getName());
      outMap.putString("displayName", pool.getDisplayName());
      return outMap;
    }, executor);
  }
}

File: ios/swift/WaasSdk/PoolService.swift
import Foundation
import WaasSdkGo
import Combine

public class PoolService: NSObject {
    // The URL of the PoolService.
    let poolServiceUrl = "https://api.developer.coinbase.com/waas/pools"

    // The handle to the Go PoolService client.
    var poolsClient: V1PoolServiceProtocol

    /**
     Initializes the PoolService with the given Cloud API Key parameters. Resolves with the string "success" on success;
     rejects with an error otherwise.
     */
    public init(_ apiKeyName: String, privateKey: String) throws {
        var error: NSError?
        let _poolsClient = V1NewPoolService(poolServiceUrl, apiKeyName, privateKey, &error)
        if (_poolsClient == nil) {
            throw WaasError.poolServiceFailedToInitialize
        }
        
        poolsClient = _poolsClient!
    }
    
    /**
     Creates a Pool with the given parameters.  Resolves with the created Pool object on success; rejects with an error
     otherwise.
     */
    public func createPool(displayName: String, poolID: String) -> Future<V1Pool, WaasError> {
        return Future() { promise in
            Job.background().async(execute: {
                do {
                    let pool = try self.poolsClient.createPool(displayName, poolID: poolID)
                    promise(Result.success(pool))
                } catch {
                    promise(Result.failure(WaasError.poolServiceUnspecifiedError(error as NSError)))
                }
            })
        }
    }
}
